Feature/seeds

cesarj45 opened the pull request:
## What kind of PR is this?

- [x] Feature

## Description

- Add a `seeds` directory and an `index.ts`.
- Add a `seed.service.ts` to call all the seeds at once.
- Add a `seed.module.ts` to manage and organize all seed-related logic.
- Add two seeds: `user.seed.ts` and `role.seed.ts`, to set the default `User (Admin)`.
- Add a `scripts/run-seed.ts` script to start the app and execute the seed service via a command.
- Modified `package.json` by adding the `seed:run` command.

## Ticket id
https://barbershop-app.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-52

![image](https://github.com/user-attachments/assets/23244dbf-41ff-4aba-92b4-f382c4380575)

cesarj45 committed:
feature: add seeds

- Add a `seeds` directory and an `index.ts`.
- Add a `seed.service.ts` to call all the seeds at once.
- Add two seeds: `user.seed.ts` and `role.seed.ts`, to set the default `User (Admin)`.
- Enable the seeds in `app.module.ts` to run automatically when the app starts.

cesarj45 committed:
add seed.service to index

cesarj45 committed:
remove role and user imports

cesarj45 committed:
create the necessary modules to use typeORM in seeds

cmagallon1 commented on src/seeds/role.seed.ts:
```diff
@@ -0,0 +1,30 @@
+
+  async run() {
+    const existingUsers = await this.roleRepository.find();
+    if (existingUsers.length > 0) {
```
💭  I wouldn't rely on the length of the existing users to know if the seeds ran, maybe there are users but not the ones expected from the seed, instead I'd use the `findOneBy` method by typeorm:

```suggestion
   const roles = [
     { name: 'Admin' },
   ]
   
   async run() {
     roles.forEach(({ name }) => {
       res = this.roleRepository.findOneBy({ name })
       if (res) { continue } // it means the role already exists
       
       // create the role
     })
   }
```

cesarj45 replied:
solved

cmagallon1 commented on src/seeds/user.seed.ts:
```diff
@@ -0,0 +1,43 @@
+
+  async run() {
+    const existingUsers = await this.userRepository.find();
+    if (existingUsers.length > 0) {
```
same here

cesarj45 replied:
solved

cmagallon1 commented on src/app.module.ts:
```diff
@@ -16,8 +23,16 @@ import { AppDataSource } from './data-source';
   ],
   controllers: [AppController],
-  providers: [AppService],
+  providers: [AppService, SeedService, RoleSeed, UserSeed],
```
just curious, why do you need the seed services in the providers lists? 🤔 is that necessary?

can't you just import the classes and execute the services when initializing the app?

cesarj45 replied:
resolved, added SeedModule

cmoreno357 replied:
@cesarj45  wait there

Hey @cmagallon1, great question! 😊

Adding SeedService, RoleSeed, and UserSeed to the providers array is necessary for a few reasons:

- Dependency Injection: NestJS relies on its DI system to manage services. By listing these services in providers, we ensure they can be injected wherever needed (e.g., in controllers or other services).

- Encapsulation and Reusability: Keeping these services in providers makes them available across the app, promoting modularity and reusability without manually instantiating them.

- Initialization Logic: If we need to run seeding logic on app startup, we can use lifecycle hooks like OnModuleInit within the service itself. This keeps the initialization logic clean and encapsulated.

Here’s an example of how SeedService could use OnModuleInit:

```typescript

import { Injectable, OnModuleInit } from '@nestjs/common';

@Injectable()
export class SeedService implements OnModuleInit {
  async onModuleInit() {
    await this.seedDatabase();
  }

  private async seedDatabase() {
    // Seeding logic here
  }
}
```
This approach aligns with NestJS best practices and keeps the app organised. For more details, check out the official docs on [Providers](https://docs.nestjs.com/providers) and [Lifecycle Events](https://docs.nestjs.com/fundamentals/lifecycle-events).

Let me know if you have further questions! 🚀

cesarj45 replied:
Yes, that's why, if the seeds are executed on the startup of the app, this is necessary, and we can avoid creating a separate module for seeds. However, @cmagallon1 requested to change that and use the seeds on demand with a command, so I decoupled the seeds from the app.module and created a new module, which is imported in the app.module anyway, but in a more orderly way

mtopete1996 replied:
No

cmagallon1 commented on src/app.module.ts:
```diff
@@ -16,8 +23,16 @@ import { AppDataSource } from './data-source';
+  constructor(private readonly seedService: SeedService) {}
+
+  async onModuleInit() {
+    await this.seedService.run();
```
💭 Additionally, I'd suggest having the seeding as a command in the `package.json` rather than when initializing the app, so you can run the seeding whenever you need it and you don't have to depend on running the app for that

cesarj45 replied:
solved

cesarj45 committed:
chore: Make the seeds more dynamic and runnable on demand.

- Add a `seed.module.ts` to manage and organize all seed-related logic.
- Add a `scripts/run-seed.ts` script to start the app and execute the seed service via a command.
- Modified `package.json` by adding the `seed:run` command.

cmoreno357 reviewed (changes requested):
Please read on the documentation within the Seed service Usage. I am addressing it in the following comment.
https://github.com/cmoreno357/mindticas2025-back/pull/21#discussion_r1982136546

mtopete1996 reviewed (changes requested):
No

mtopete1996 reviewed (approved):
Ok it does work
